### PR TITLE
Feature/url builder update

### PR DESCRIPTION
**Pull request by cvasqu09**
Refactored the UrlBuilder and changed categories to only include those which will return at least 10 questions for the combinations of game length and difficulty.

### CODE REVIEW

File: core/src/main/java/GameObjects/UrlBuilderImp.java [hunk 1]
```diff
@@ -1,22 +1,61 @@
...
+      return "22";
+    } else if (category.compareTo(Lobby.MUSICTEXT) == 0) {
+      return "12";
+    } else if (category.compareTo(Lobby.VIDEOGAMESTEXT) == 0) {
+      return "15";
+    } else if (category.compareTo(Lobby.SCIENCENATURETEXT) == 0) {
+      return "17";
+    } else {
+      return "23";
+    }
```
Comment (jmtaber129) [inline core/src/main/java/GameObjects/UrlBuilderImp.java:24]: I know you can switch on string literals, but I'm not sure about strings that are static final.  Use a switch if it works.

File: core/src/main/java/Scenes/Lobby.java [hunk 2]
```diff
@@ -41,12 +40,25 @@
...
+  public static final String MUSICTEXT = "Music";
+  public static final String VIDEOGAMESTEXT = "Video Games";
+  public static final String SCIENCENATURETEXT = "Science and Nature";
+  public static final String HISTORYTEXT = "History";
+  public static final String EASYTEXT = "Easy";
+  public static final String MEDIUMTEXT = "Medium";
+  public static final String HARDTEXT = "Hard";
+  public static final String SHORTTEXT = "Short (5)";
+  public static final String MEDIUMLENGTHTEXT = "Medium (7)";
+  public static final String LONGTEXT = "Long (9)";
```
Comment (jmtaber129) [inline core/src/main/java/Scenes/Lobby.java]: Use underscores to separate words, e.g. "GENERAL_TEXT".

File: core/src/main/java/Scenes/Lobby.java [hunk 3]
```diff
@@ -41,12 +40,25 @@
...
+  public static final String GEOGRAPHYTEXT = "Geography";
+  public static final String MUSICTEXT = "Music";
+  public static final String VIDEOGAMESTEXT = "Video Games";
+  public static final String SCIENCENATURETEXT = "Science and Nature";
+  public static final String HISTORYTEXT = "History";
+  public static final String EASYTEXT = "Easy";
+  public static final String MEDIUMTEXT = "Medium";
+  public static final String HARDTEXT = "Hard";
+  public static final String SHORTTEXT = "Short (5)";
+  public static final String MEDIUMLENGTHTEXT = "Medium (7)";
```
Comment (jmtaber129) [inline core/src/main/java/Scenes/Lobby.java]: Just say "MEDIUM_TEXT" since the short and long options don't include "LENGTH".

**Review by jmtaber129 — CHANGES_REQUESTED** [review]
Now that I see how passing settings as strings is implemented, I think it might be better to create a GameSettings object that holds the selected game options, as well as the constants used as the button text.  This implementation would also allow the get<Option>Value() methods to be implemented in a class that already holds the relevant information.

File: core/src/main/java/Scenes/Lobby.java (hunk 3 above)
Comment (cvasqu09) [reply core/src/main/java/Scenes/Lobby.java]: I am going to change it to NORMAL. The reason that I have LENGTH at the end is because we have a difficulty and a gameplay length with the name MEDIUM.

File: core/src/main/java/GameObjects/UrlBuilderImp.java (hunk 1 above)
Comment (cvasqu09) [reply core/src/main/java/GameObjects/UrlBuilderImp.java:24]: I believed I tried using a case statement and was getting an error but I'll check again.

Comment (cvasqu09) [reply core/src/main/java/GameObjects/UrlBuilderImp.java:24]: According to Oracle the switch statement allows for strings in Java 7 and up however I still get an error when trying to use the switch inside of Android Studio.

File: core/src/main/java/GameObjects/GameSettings.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      return "22";
+    } else if (category.compareTo(Lobby.MUSIC_TEXT) == 0) {
+      return "12";
+    } else if (category.compareTo(Lobby.VIDEOGAMES_TEXT) == 0) {
+      return "15";
+    } else if (category.compareTo(Lobby.SCIENCE_NATURE_TEXT) == 0) {
+      return "17";
+    } else {
+      return "23";
+    }
```
Comment (jmtaber129) [inline core/src/main/java/GameObjects/GameSettings.java:37]: Didn't think about this earlier, but I'm pretty sure there's an equals method for strings that would be better for checking equality (something like string1.equals(string2)).

File: core/src/main/java/GameObjects/GameSettings.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  public String getCategoryValue() {
+    if (category.compareTo(Lobby.GENERAL_TEXT) == 0) {
+      return "9";
+    } else if (category.compareTo(Lobby.GEOGRAPHY_TEXT) == 0) {
+      return "22";
+    } else if (category.compareTo(Lobby.MUSIC_TEXT) == 0) {
+      return "12";
+    } else if (category.compareTo(Lobby.VIDEOGAMES_TEXT) == 0) {
+      return "15";
+    } else if (category.compareTo(Lobby.SCIENCE_NATURE_TEXT) == 0) {
```
Comment (jmtaber129) [inline core/src/main/java/GameObjects/GameSettings.java]: It might be better to move Lobby.SETTING_TEXT into this class, and have Lobby get these constants from this class.

File: core/src/main/java/GameObjects/UrlBuilder.java [hunk 6]
```diff
@@ -5,5 +5,5 @@
  */
 
 public interface UrlBuilder {
-  String buildUrl(String amount, String category, String difficulty);
+  String buildUrl();
```
Comment (jmtaber129) [inline core/src/main/java/GameObjects/UrlBuilder.java]: buildUrl should now be taking in a GameSettings object, rather than creating a new UrlBuilder for each settings object (Essentially replacing the separate string parameters with one object that holds the strings).

File: core/src/main/java/GameObjects/UrlBuilderImp.java [hunk 7]
```diff
@@ -12,50 +12,16 @@
   private static final String CATEGORY = "category=";
   private static final String AMOUNT = "amount=";
   private static final String MULTIPLE_CHOICE_QUERY = "type=multiple";
-
-  public UrlBuilderImp() {
+  private static GameSettings settings;
+  
+  public UrlBuilderImp(GameSettings settings) {
+    this.settings = settings;
```
Comment (jmtaber129) [inline core/src/main/java/GameObjects/UrlBuilderImp.java]: See my comment in UrlBuilder.java

File: core/src/main/java/GameObjects/GameSettings.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
+package GameObjects;
+
+import Scenes.Lobby;
+
+/**
+ * The GameSettings class will store the information for the preferences that were selected in the
+ * lobby screen and will have methods that return values corresponding to the values needed in the
+ * buildUrl method of the UrlBuilderImp class.
+ */
+public class GameSettings {
```
Comment (jmtaber129) [inline core/src/main/java/GameObjects/GameSettings.java:10]: Add unit tests for this class.  Especially now that most of the logic in obtaining the API representation of each of the settings is implemented in this class.

File: core/src/main/java/Scenes/Lobby.java (hunk 3 above)
Comment (cvasqu09) [reply core/src/main/java/Scenes/Lobby.java]: Fixed

File: core/src/main/java/Scenes/Lobby.java (hunk 2 above)
Comment (cvasqu09) [reply core/src/main/java/Scenes/Lobby.java]: Fixed

File: core/src/test/java/GameObjects/UrlBuilderImpTest.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for the UrlBuilder. The URLs provided were taken directly from the database using their
+ * API.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class UrlBuilderImpTest {
```
Comment (jmtaber129) [inline core/src/test/java/GameObjects/UrlBuilderImpTest.java:17]: A lot of these tests test basically the same thing.  I.e., each of these test that the urlBuilder takes the return value of the settings object, and puts it in the correct place in the url.  Consider removing a few of the redundant tests, and adding some tests where the settings object returns an unexpected value, like "foo", an empty string, or null.

File: core/src/main/java/GameObjects/GameSettings.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      return "medium";
+    } else {
+      return "hard";
+    }
+  }
+
+  public String getAmountValue() {
+    if (amount.compareTo("Short (5)") == 0) {
+      return "5";
+    } else if (amount.compareTo("Medium (7)") == 0) {
```
Comment (jmtaber129) [inline core/src/main/java/GameObjects/GameSettings.java]: Use SHORT_TEXT and MEDIUM_TEXT.

File: core/src/test/java/GameObjects/UrlBuilderImpTest.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+  GameSettings settingsMock;
+
+  @InjectMocks
+  UrlBuilderImp urlBuilder;
+
+  @Test
+  /**
+   * Test when general category is used with 9 questions and easy difficulty.
+   */
+  public void longGeneralEasy(){
```
Comment (jmtaber129) [inline core/src/test/java/GameObjects/UrlBuilderImpTest.java]: I'd prefer test method names of the form <methodName>_<state>() for consistency with the file management tests.  E.g. "buildUrl_longGeneralEasy()".  However, the style guide suggests test<methodName>_<state>() (probably to avoid confusion with test helper methods), so we might change it in the future.

**Review by jmtaber129 — CHANGES_REQUESTED** [review]

File: core/src/main/java/GameObjects/GameSettings.java (hunk 5 above)
Comment (cvasqu09) [reply core/src/main/java/GameObjects/GameSettings.java]: I disagree. It doesn't seem natural to move all the Strings from the Lobby to the GameSettings class. If we want to change the text of the buttons for Lobby we would have to go to the GameSettings class to do so.

File: core/src/test/java/GameObjects/UrlBuilderImpTest.java (hunk 9 above)
Comment (cvasqu09) [reply core/src/test/java/GameObjects/UrlBuilderImpTest.java:17]: What do we want the urlBuilder to throw in the case that it receives these unexpected values?

File: core/src/main/java/GameObjects/GameSettings.java (hunk 5 above)
Comment (jmtaber129) [reply core/src/main/java/GameObjects/GameSettings.java]: I think it would be reasonable, since these constants are the textual representations of the settings, and the buttons display this textual representation.  I'll leave it up to you.

File: core/src/main/java/GameObjects/UrlBuilderImp.java (hunk 1 above)
Comment (jmtaber129) [reply core/src/main/java/GameObjects/UrlBuilderImp.java:24]: From what I was looking at, it seems like strings in switches only work with string literals, due to the way the compiler deals with them.

File: core/src/test/java/GameObjects/UrlBuilderImpTest.java (hunk 9 above)
Comment (jmtaber129) [reply core/src/test/java/GameObjects/UrlBuilderImpTest.java:17]: I think the "foo" example would just use "foo" as the value.  For empty or null strings, I think urlBuilder should either return a null or empty string for the entire method to denote invalid settings, or throw an IllegalArgumentException.

File: core/src/test/java/GameObjects/GameSettingsTest.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
+package GameObjects;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Created by Chris on 11/20/2016.
+ */
+public class GameSettingsTest {
```
Comment (jmtaber129) [inline core/src/test/java/GameObjects/GameSettingsTest.java:12]: Add tests for invalid constructor arguments.

**Review by jmtaber129 — CHANGES_REQUESTED** [review]

File: core/src/main/java/GameObjects/GameSettings.java (hunk 8 above)
Comment (cvasqu09) [reply core/src/main/java/GameObjects/GameSettings.java:10]: Fixed

File: core/src/test/java/GameObjects/UrlBuilderImpTest.java (hunk 9 above)
Comment (cvasqu09) [reply core/src/test/java/GameObjects/UrlBuilderImpTest.java:17]: Fixed

File: core/src/test/java/GameObjects/GameSettingsTest.java (hunk 12 above)
Comment (cvasqu09) [reply core/src/test/java/GameObjects/GameSettingsTest.java:12]: Fixed

**Review by jmtaber129 — APPROVED** [review]
